File: NextGreaterPermutation/NextGreaterPermutation.cpp
// NextGreaterPermutation.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>
#include <algorithm>

//void TraceVector(const std::vector<int>& v)
//{
//    std::cout << "{";
//    for (auto i = v.cbegin(); i != v.cend(); ++i)
//    {
//        if (i != v.cbegin())
//            std::cout << ", ";
//        std::cout << *i;
//    }
//
//    std::cout << "}" << std::endl;
//}
//
void AddPermutationNoDuplicates(std::vector<std::vector<int>>& arrays, const std::vector<int>& arr)
{
    for (auto& ar : arrays)
    {
        bool arrayFound = true;
        for (size_t i = 0; i < arr.size(); ++i)
        {
            if (ar[i] != arr[i])
            {
                arrayFound = false;
                break;
            }
        }
        if (arrayFound)
            return;
    }

    arrays.push_back(arr);
}

//int tried = 0;
void Permutations1(std::vector<std::vector<int>>& permutations, std::vector<int>& arr, int index)
{
    //TraceVector(arr);
    //tried++;
    AddPermutationNoDuplicates(permutations, arr);

    for (size_t i = index; i < arr.size() - 1; ++i)
    {
        for (size_t j = i + 1; j < arr.size(); ++j)
        {
            std::swap(arr[i], arr[j]);
            Permutations1(permutations, arr, index + 1);
            std::swap(arr[i], arr[j]);
        }
    }
}

void Permutations(std::vector<std::vector<int>>& permutations, std::vector<int>& arr, int index)
{
    //tried++;
    if (index == arr.size() - 1)
    {
        AddPermutationNoDuplicates(permutations, arr);
        return;
    }

    for (size_t i = index; i < arr.size(); ++i)
    {
        std::swap(arr[i], arr[index]);
        Permutations(permutations, arr, index+1);
        std::swap(arr[i], arr[index]);
    }
}

void NextGreaterPermutation(std::vector<int>& arr)
{
    std::vector<std::vector<int>> permutations;
    
    Permutations(permutations, arr, 0);
    //std::cout << "Tried " << tried << " times" << std::endl;
    std::sort(permutations.begin(), permutations.end());

    for (auto i = permutations.cbegin(); i != permutations.cend(); ++i)
    {
        if (arr == *i)
        {
           arr.clear();
           if (i + 1 == permutations.cend())
           {
               std::vector<int> v(*permutations.cbegin());
               arr.assign(v.cbegin(), v.cend());
           }
           else
           {
               std::vector<int> v(*(i + 1));
               arr.assign(v.cbegin(), v.cend());
           }
           return;
        }
    }
}

int main()
{
    while (true)
    {
        //tried = 0;
        int count = 0;
        std::cout << "Number of elements in the array: ";
        std::cin >> count;
        if (count == 0)
            break;

        std::vector<int> vec(count);
        std::cout << "The number array: ";
        for (int i = 0; i < count; ++i)
            std::cin >> vec[i];

        std::cout << "Next Greater Permutation: " << std::endl;
        NextGreaterPermutation(vec);
        for (size_t i = 0; i < vec.size(); ++i)
        {
            if (i != 0)
                std::cout << ", ";
            std::cout << vec[i];
        }

        std::cout << std::endl;
    }
}

File: Go/NextGreaterPermutation.go
package main

import (
	"fmt"
	"sort"
)

func PermutationsEqual(a []int, b []int) bool {
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}

	return true
}

func PermutationExists(vecs [][]int, vec []int) bool {
	for i := range vecs {
		if PermutationsEqual(vecs[i], vec) {
			return true
		}
	}

	return false
}

func GetPermutations(arr []int, index int, permutations *[][]int) {
	if index == len(arr)-1 {
		if PermutationExists(*permutations, arr) {
			return
		}

		temp := make([]int, len(arr))
		copy(temp, arr)
		*permutations = append(*permutations, temp)
		return
	}

	for i := index; i < len(arr); i++ {
		arr[index], arr[i] = arr[i], arr[index]
		GetPermutations(arr, index+1, permutations)
		arr[index], arr[i] = arr[i], arr[index]
	}
}

func NextGreaterPermutation(arr []int) {
	var permutations [][]int

	GetPermutations(arr, 0, &permutations)

	sort.Slice(permutations, func(i, j int) bool {
		for k := range permutations[i] {
			if permutations[i][k] != permutations[j][k] {
				return permutations[i][k] < permutations[j][k]
			}
		}
		return false
	})

	for i := 0; i < len(permutations); i++ {
		if PermutationsEqual(arr, permutations[i]) {
			i++
			if i == len(permutations) {
				i = 0
			}
			copy(arr, permutations[i])
			break
		}
	}
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		count := 0
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := make([]int, count)
		fmt.Print("Please enter the array: ")
		for i := 0; i < count; i++ {
			fmt.Scan(&vec[i])
		}

		NextGreaterPermutation(vec)

		fmt.Print("The next permutation is ")
		fmt.Println(vec)
	}
}
